[TypeInfo] Redesign Type methods and nullability

mtarld opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        |
| License       | MIT

By using the component and trying to integrate it within several libs, I ended up finding some limitations due to the component design.

Basically, the issues were:
- Abstract classes (such as `Type`) knowing its concrete implementations (such as in `getBaseType` method).
- For API simplicity `CollectionType`, and `GenericType` are forwarding methods to their wrapped type using `__call`, but I think this must be avoided as much as possible.
- It is difficult to know if a type is "just" a nullable type (because we need to check the union content).
- Invalid types could be instantiated

To tackle these issues, I propose the following:
- Introduce `CompositeTypeInterface` and `WrappingTypeInterface` that describe whether a type is composed by several types or wrapping one.
- Add a `NullableType` which extends `UnionType` and is a `WrappingTypeInterface`, so that it is easy to recognize a nullable type and to get the non-nullable related type.
- Remove the `Type::asNonNullable` method as it's covered by `NullableType::getWrappedType`
- Get rid of the `__call` method as it is easy to know it a type is wrapping another one
- Rename `is` to `satisfy` and add the new `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` methods
- Rename `isA` to `isIdentifiedBy`

Plus, **thanks to @skalpa's great work**, the following checks are added to prevent invalid type construction:
- `BackedEnumType`
    - Backed type can only be `int` or `string`
- `CollectionType`
    - Main type can only be `array` or `iterable` when it's a built-in type
- `GenericType`
    - Main type can only be `array` or `iterable` when it's a built-in type
- `IntersectionType`
    - Cannot contain other types than class types
- `UnionType`
    - Cannot contain `UnionType`
    - Cannot contain standalone types (`mixed`, `void`, and `never`)
    - Cannot contain either `true` and `false`
    - Cannot contain either `bool` and `true` or `false`
    - Cannot contain either `object` and a class type

stof commented:
Using public properties instead of getters makes the maintenance harder in case we need to make the API of the component evolve (as we cannot rely on property hooks yet to implement BC layers)

stof commented on src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php:
```diff
@@ -11,6 +11,7 @@
 
 namespace Symfony\Component\PropertyInfo\Extractor;
 
+use Symfony\Component\TypeInfo\Specification\IsNullable;
```
this seems unused

stof commented on src/Symfony/Component/TypeInfo/Specification/IsA.php:
```diff
@@ -0,0 +1,67 @@
+    ) {
+    }
+
+    public function isSatifiedBy(Type $type): bool
```
There is a big drawback of this proposal: the implementation of this method assumes that it knows all the possible implementations of Type (as it has to handle them in different way). This assumption is wrong (nothing forbids someone to create another class extending Type)

mtarld replied:
Indeed, you were totally right, I moved back to the "abstract class" methods but in a way that ignores the concrete implementations.

jdreesen commented on src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php:
```diff
@@ -128,7 +128,9 @@ public function getType(DocType $varType): ?Type
             }
 
-            return $this->createType($varType, $nullable);
+            $type = $this->createType($varType, $nullable);
```
```suggestion
            $type = $this->createType($varType);
```

mtarld replied:
Nice catch! Thanks :slightly_smiling_face:

Korbeil reviewed (approved):
Loving all the changes ! :rocket:

xabbuh commented on src/Symfony/Bridge/Doctrine/composer.json:
```diff
@@ -39,7 +39,7 @@
         "symfony/stopwatch": "^6.4|^7.0",
         "symfony/translation": "^6.4|^7.0",
-        "symfony/type-info": "^7.1",
+        "symfony/type-info": "^7.2",
```
Is this version bump needed?

mtarld replied:
Indeed, it wasn't, thanks!

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -64,7 +64,7 @@
         "symfony/translation": "^6.4|^7.0",
         "symfony/twig-bundle": "^6.4|^7.0",
-        "symfony/type-info": "^7.1",
+        "symfony/type-info": "^7.2",
```
Is this version bump needed?

xabbuh commented on src/Symfony/Component/TypeInfo/Type/CompositeTypeInterface.php:
```diff
@@ -0,0 +1,31 @@
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * Represents a type composed by several other types.
```
```suggestion
 * Represents a type composed of several other types.
```

xabbuh commented on src/Symfony/Component/TypeInfo/Type/TemplateType.php:
```diff
@@ -27,33 +25,18 @@ final class TemplateType extends Type
-    }
-
-    public function asNonNullable(): self
+    public function getBoundType(): Type
```
the renaming from `getBound()` to `getBoundType()` should also be documented in the changelog file

stof commented on src/Symfony/Component/TypeInfo/Type.php:
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
+    /**
+     * Tells if the type or one of its component (@see traverse()) is nullable.
+     *
+     * @param TypeIdentifier|class-string $expected
```
wrong phpdoc

stof commented on src/Symfony/Component/TypeInfo/Type.php:
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
+    public function isScalar(): bool
     {
-        throw new LogicException(\sprintf('Cannot call "%s" on "%s" type.', $method, $this));
+        return $this->isA(TypeIdentifier::INT)
```
instead of separate `isA` calls that will require separate traversals, we should perform a single traversing using a combined check for all identifiers (maybe with an API inspired by `is` taking a callback with a type)

mtarld replied:
Great idea yes, I updated the method signature so that it can take a variadic argument

stof commented on src/Symfony/Component/TypeInfo/Type.php:
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
     {
-        return $callable($this);
+        foreach ($this->traverse() as $type) {
+            if ($type !== $this && $type->isA($expected)) {
```
Wouldn't that perform multiple traversals of parts of the object graph when one of the nested type is itself a composite or a wrapping type, and this type does not override `isA` ? `traverse` does not yield only leaf nodes

stof commented on src/Symfony/Component/TypeInfo/Type/NullableType.php:
```diff
@@ -0,0 +1,48 @@
+        private readonly Type $type,
+    ) {
+        if ($type instanceof UnionType) {
+            parent::__construct(Type::null(), ...$type->getTypes());
```
should we really perform this unwrapping here ?

mtarld replied:
As `UnionType` has a variadic argument, yes I think

stof commented on src/Symfony/Component/TypeInfo/Type/NullableType.php:
```diff
@@ -0,0 +1,48 @@
+        parent::__construct(Type::null(), $type);
+    }
+
+    public function getWrappedType(): Type
```
this type actually composes both this type and the NULL type. It might cause confusion for code not handling properly types implementing both WrappingTypeInterface and IntersectionType.

And this `getWrappedType` is not even a replacement for `asNonNullable` as nothing prevents the NULL type to be part of the wrapped type.

mtarld replied:
> this type actually composes both this type and the NULL type. It might cause confusion for code not handling properly types implementing both WrappingTypeInterface and IntersectionType.

I'm not sure to understand what you mean here, could you elaborate please?

> And this `getWrappedType` is not even a replacement for `asNonNullable` as nothing prevents the NULL type to be part of the wrapped type.

From now on, the `NullableType` can only be constructed with a non-nullable type, therefore `asNonNullable` is now equal to `getWrappedType`.

stof commented on src/Symfony/Component/TypeInfo/Type.php:
```diff
@@ -25,35 +24,70 @@ abstract class Type implements \Stringable
 
     /**
-     * @param callable(Type): bool $callable
+     * Tells if the type or one of its component (@see traverse()) is what is expected.
```
This `isA` API is confusing IMO, because it does not allow to determine whether a type is either a subtype nor a supertype of another type.
Common questions related to types would be "does this type accept this kind of value?", which cannot be answered by this API (the best proof is the fact that both UnionType and IntersectionType uses this same implementation in the base class).
The `is` method that you removed in this PR *was* able to answer such question.

skalpa replied:
@stof. Agreed, that is exactly one of my use case:  I need to determine whether a particular value is valid where a specific type-hint is used and this PR would render this method useless to me.
However, the current implementation has issues nevertheless: on one side `Type::object('SomeClass')->isA('SomeInterface')` but on the other `Type::union(Type::int(), Type::string())->isA(TypeIdentifier::INT)`: it's not clear whether the type is supposed to accept what is passed to isA() or if it is the other way around. 
I was actually going to work on type comparisons, I'd suggest leaving `isA()` untouched for now and reviewing/improving the current implementation separately.

stof replied:
Well, I totally agree that `isA` is weird (maybe it only needs to be renamed). But that's one more reason why removing `is` by suggesting `isA` as a replacement is a very bad idea.

mtarld replied:
You're totally right, that's why I renamed `Type::is` to `Type::satisfy` and `Type::isA` to `Type::isIdentifiedBy` so that it is easier to understand. And I added the `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` so that it is easier to define whether we want to check the root, the children or the leaves.

stof commented on src/Symfony/Component/TypeInfo/Type/BuiltinType.php:
```diff
@@ -46,43 +40,13 @@ public function getTypeIdentifier(): TypeIdentifier
-        }
-
-        // "mixed" is an alias of "object|resource|array|string|float|int|bool|null"
-        // therefore, its non-nullable version is "object|resource|array|string|float|int|bool"
```
the removal of this logic is another proof that your assumption that making a non-nullable is simple without a dedicated API is wrong.

stof commented on src/Symfony/Component/TypeInfo/Type/TemplateType.php:
```diff
@@ -27,33 +25,18 @@ final class TemplateType extends Type
         return $this->name;
     }
 
-    public function getBound(): Type
```
is the renaming of this method really worth a BC break ? Experimental components allow BC breaks outside major versions, but we still try to maintain BC as much as possible to make upgrades easier.

mtarld replied:
I didn't know that we had to be that careful with experimental components, I reverted that, thanks

stof reviewed (commented):
In the current state of this PR, I'm voting -1 as it reduces the power of the component.

skalpa commented:
> It is difficult to know if a type is "just" a nullable type (because we need to check the union content).

I am not sure I understand what problem your are trying to solve. A type is "just" a nullable type if `$type instanceof BuiltinType && $type->isA(TypeIdentifier::NULL)`, or you could just cast your type to string ? I.e. `'null' === (string) $type`.

Nevertheless, the changes feel a lot complex considering nullability is not a generic problem. There are only 3 kinds of nullable types: `null`, `mixed`, and unions that contain `null`.

Therefore, I would only keep the current `isNullable()`, and `asNonNullable()` methods, but refactor them a bit:

* The implementation of `Type::isNullable()` should just return `false`
* `Type::asNonNullable()` should be implemented and just return `$this`
* All the unnecessary `asNonNullable()` in children classes can be removed
* `BuiltinType` has to override `isNullable()` and return `true` when appropriate

With these little changes I believe nullability would be properly covered.

stof commented:
@skalpa actually, there are more nullable types:

- an intersection type where all its types are nullable
- a union type containing a nullable type (which might not be `null`, as you said yourselves that there are more nullable types, for instance `mixed` or other union types)

skalpa commented:
> @skalpa actually, there are more nullable types:
> * an intersection type where all its types are nullable
> * a union type containing a nullable type (which might not be `null`, as you said yourselves that there are more nullable types, for instance `mixed` or other union types)

@stof, intersections cannot contain nullable types, only [class/interface names](https://www.php.net/manual/en/language.types.type-system.php), which is logical (nothing could be both an instance of `Stuff` and `null` at the same time). Currently it is possible to create nonsensical intersections that would be invalid in PHP, but I fixed that in #57943.

For unions, yes, I meant "a union that contains a nullable" and not "a union that contains null", but after seeing your comment and checking, I believe it should be forbidden as well. Trying to add `mixed` to a union gives me `PHP Fatal error:  Type mixed can only be used as a standalone type`.

And a union cannot contain another union: this throws an exception (`Type::union()` lets you pass other unions but it merges them to a new one before calling the constructor).

stof commented:
@skalpa as said above, `mixed` is nullable. And it could be used in an IntersectionType (unless you canonicalize such intersection `mixed&T` to be `T` during instantiation).

And why the PHP syntax enforces types to be written in disjunctive normal form (meaning no union inside intersection), I'm not sure TypeInfo should have this restriction on the created object graph (or it should have proper types enforcing valid usages).
And enforcing canonical forms might make `asNonNullable` more complex because of the fact that turning `mixed` into a non-nullable type turns it into a union.

skalpa commented:
> @skalpa as said above, `mixed` is nullable. And it could be used in an IntersectionType (unless you canonicalize such intersection `mixed&T` to be `T` during instantiation).
> 

@stof, No, an intersection cannot contain `mixed`. I am sorry to insist, but this is absolutely crucial as it completely changes the semantics of the type if not enforced. In PHP, intersections represent objects. If a value satisfies an intersection, then one must be able to assume that `true === is_object($value)`. And PHP forbids using anything else than class names in intersections.

While the normal form is something else, I really believe we should prevent the creation of nonsensical and/or invalid types that would never be satisfied by any value or would only result in PHP fatal errors if used, as they may probably be indicators of a bug in user code.

If this is not enforced by the component, then as a user I will have to anticipate absolutely impossible cases just because the API allows them and I'd have to protect my code against intersections of scalars, or unions of `mixed` and `never`. ☹️

(even though yes, technically you're right, any intersection is always equivalent to itself & mixed)

> And why the PHP syntax enforces types to be written in disjunctive normal form (meaning no union inside intersection), I'm not sure TypeInfo should have this restriction on the created object graph (or it should have proper types enforcing valid usages). And enforcing canonical forms might make `asNonNullable` more complex because of the fact that turning `mixed` into a non-nullable type turns it into a union.

Some restrictions/enforcements are already there: you are not allowed to add a union to a union, or an intersection to another intersection, which is completely arbitrary if you consider that these could be merged later. But in fact I believe that @mtarld's is the right approach here and am just trying to build on top of it:

IMHO, object graphs should be as canonical as possible. For instance, the fact that unions sort their subtypes is brilliant and without it, to compare `Type::union(Type::int(), Type::string())` to `Type::union(Type::string(), Type::int())` would be near impossible. 

To make this possible requires enforcing the normal form in the `Type` classes afaik. This is actually partly implemented but incomplete, which is what I tried to improve in my own PR.

In return, `TypeFactoryTrait` methods can be used to add some flexibility to types creation and ensure that `Type::union(Type::int(), Type::union(Type::string(), Type::bool()))` is not only allowed but will create the exact same type as `Type::union(Type::bool(), Type::int(), Type::string())`.

If we consider that we should allow any random combination of types, then the current checks should be removed (but I would consider that a step back).

TL;DR (sorry but I really like this component, I talk too much sometimes): I'd like to be able to assume that `echo $type` will always give me a syntactically valid PHP type. We can already use `Type::union()` and `Type::intersection()` to combine composite types in a more flexible way and could add more if necessary.

mtarld commented:
Ok, I ended up with the following:
- I renamed `Type::is` to `Type::satisfy` and `Type::isA` to `Type::isIdentifiedBy` so that it is easier to understand
- I added the `CompositeTypeInterface::composedTypesSatisfy` and `WrappingTypeInterface::wrappedTypeSatisfy` methods so that it is easier to define whether we want to check the root, the children or the leaves (you can have a look for an example in `Type::isIdentifiedBy`

I'm still not sure if the `Type::asNonNullable` is needed, as we now have that `NullableType`. I mean, does it make sense to make `mixed` or `null` as non nullable? To me, the main aim for that method me is to recover `int` in `int|null`, which is covered by `NullableType::getWrappedType`.

WDYT?

mtarld committed:
[TypeInfo] Redesign Type methods and nullability

mtarld commented:
Moreover, I pulled some ideas from the great work of @skalpa in https://github.com/symfony/symfony/pull/57943 that make types more consistent and make my work on nullable types more easy and relevant.

chalasr commented:
Rebase needed.
Also we need to move forward on this as the 7.2 feature freeze will come fast making TypeInfo's API stable and there are several open PRs depend on the API that is changed here. @stof can you please have another look?

xabbuh commented:
The failures look related.

mtarld commented:
These changes are making types generation stricter, and they allowed us to catch a bug: https://github.com/symfony/symfony/pull/58371

chalasr commented on src/Symfony/Component/TypeInfo/Type.php:
```diff
@@ -25,35 +24,36 @@ abstract class Type implements \Stringable
-    abstract public function isA(TypeIdentifier|string $subject): bool;
-
-    abstract public function asNonNullable(): self;
+    public function satisfy(callable $specification): bool
```
`isSatisfiedBy()` instead?

chalasr commented on src/Symfony/Component/TypeInfo/CHANGELOG.md:
```diff
@@ -4,6 +4,13 @@ CHANGELOG
+ * Add `TypeIdentifier::isStandalone`, `TypeIdentifier::isScalar`, and `TypeIdentifier::isBool` methods
+ * Add `WrappingTypeInterface` and `CompositeTypeInterface` type interfaces
+ * Add `NullableType` type class
+ * Rename `Type::isA` to `Type::isIdentifiedBy` and `Type::is` to `Type::satisfy`
```
```suggestion
 * Rename `Type::isA()` to `Type::isIdentifiedBy()` and `Type::is()` to `Type::isSatisfiedBy()`
```
same for other referenced methods about the `()` suffix

chalasr reviewed (changes requested):
Please add some phpdoc to `isidentifiedBy()` and `isSatisfiedBy()` explaining what they are supposed to do and how they differ, so that one doesn't have to read the implementation for that

chalasr commented on src/Symfony/Component/TypeInfo/CHANGELOG.md:
```diff
@@ -4,6 +4,13 @@ CHANGELOG
+ * Add `TypeIdentifier::isStandalone`, `TypeIdentifier::isScalar`, and `TypeIdentifier::isBool` methods
+ * Add `WrappingTypeInterface` and `CompositeTypeInterface` type interfaces
+ * Add `NullableType` type class
+ * Rename `Type::isA` to `Type::isIdentifiedBy` and `Type::is` to `Type::isSatisfiedBy`
+ * Remove `Type::getBaseType`, `Type::asNonNullable` and `Type::__call` methods
```
```suggestion
 * Rename `Type::isA()` to `Type::isIdentifiedBy()` and `Type::is()` to `Type::isSatisfiedBy()`
 * Remove `Type::getBaseType()`, `Type::asNonNullable()` and `Type::__call()` methods
```

chalasr reviewed (approved):
Better API

mtarld commented:
The high-deps build is failing because this PR purposely contains BC breaks as the component is experimental in 7.2. Failures will go away once this is merged.

chalasr commented:
Indeed, tests will pass once merged. Then dependant components' tests will start failing on 7.1 as they will install type-info 7.2 so we will have to forbid it by adding conflict rules.

xabbuh commented:
The issue is that the conflict will lead to the fact that Composer may not install the latest 7.1 version of said component because of the conflict when 7.2 of the TypeInfo component is installed as we already have releases claiming to be compatible with TypeInfo 7.2. So the damage is already done. The only solution I see is to make the PropertyInfo component compatible with TypeInfo 7.2 as a bugfix in 7.1.

In the future we need to be careful when adding requirements for experimental components and do not allow to install new minor versions of the from the beginning.

stof commented:
@xabbuh a way to avoid that breakage is to make type-info conflict with older versions of those components, to workaround the fact that we used semver constraints for an experimental package not guaranteeing semver yet.

chalasr commented:
Sounds good

xabbuh commented:
Indeed, @stof's suggestion is a good idea that we can use in this PR.

mtarld commented:
Yes, that a great idea, I just pushed that fix.

nicolas-grekas reviewed (approved):
Let's give this a try :)

nicolas-grekas commented:
Thank you @mtarld.
